api/v1alpha1/kubevirtcluster_types.go

```diff
@@ -59,6 +59,11 @@ type KubevirtClusterSpec struct {
 	// InfraClusterSecretRef is a reference to a secret with a kubeconfig for external cluster used for infra.
 	// +optional
 	InfraClusterSecretRef *corev1.ObjectReference `json:"infraClusterSecretRef,omitempty"`
+
+	// CloudProvider specify which cloud-provider-kubevirt manager version to install at infra cluster, if not
+	// specified it will not be installed.
+	// +optional
+	CloudProvider *CloudProvider `json:"cloudProvider,omitempty"`
```

Comment:
I went back and forth on this a bit, but I think we should reconsider having the KubeVirtCluster object manage installing the cloud provider.

Part of that reasoning is that I think that the cloud provider version in use is more closely tied to the guest cluster version than the capk controller. If we put the cloud provider install in capk, then we're also making our controller in charge of the lifecycle management of that component, which means managing things like updates of the cloud provider.

I want to reach a workflow where the cloud provider can be installed declaratively along side of the capi cluster without having to serialize the logic though... meaning we should be able to post the capk cluster + cloud provider install all in one go and not worry about the timing of waiting for guest cluster kubeconfigs to get created and stuff like that.

Is there a way we can publish a templated yaml file for installing cloud-provider-kubevirt that simply takes something like a cluster name and namespace as input which allows us install cloud-provider-kubevirt easily with a guest cluster?

Maybe the workflow could look something like this...

* clusterctl generate cluster <cluster name> ... > tenant-cluster.yaml
* clusterkubevirtadm generate kccm <cluster name> ... > tenant-kccm.yaml
* kubectl apply -f tenant-cluster.yaml
* kubectl apply -f tenant-kccm.yaml

In this theoretical example,  clusterkubevirtadm tool (which is in this capk repo) would pull a templated cloud-provider-kubevirt manifest that we publish in the cloud-provider-kubevirt repo and fill in the clustername, namespace, etc... then spit out some yaml we can use to install the cloud provider along side the tenant-cluster in the infra... basically it would be the same relationship between clusterctl and our templated capk manifests.

Would that make sense to you all?

Reply:
I think kccm does not fit into clusterkubevirtadm since this tool is for the external infra cluster scenario, this is the flow from what I understand, we have the admin and the user (It will be able to create the cluster a infra cluster and that's all)
- admin -> create namespace at external infra cluster for the new tenant
- admin -> call clusterkubevirtadm that generates credentials at external infra cluster
- admin -> create secret at managed cluster with the external infra cluster kubeconfig and pass secret name to user
- user -> create a cluster template pointing to the infra cluster secret call clusterctl to generate the cluster
- user -> apply the generated cluster that will create the tenant on external infra.

So installing kccm is adding a feature to the cluster and has to be something decided by the user not the admin so it fits on clusterctl and not at clusterkubevirtadm.

Declarativeness the proposed solution is is quite declarative I would say
```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha1                            
kind: KubevirtCluster                                                           
metadata:                                                                       
  name: "${CLUSTER_NAME}"                                                       
  namespace: "${NAMESPACE}"                                                     
spec:                                                                           
  cloudProvider:                                                                
    loadBalancer:                                                               
      enabled: true                                                             
      creationPollInterval: 30                                                  
  controlPlaneServiceTemplate:                                                  
    spec:                                                                       
      type: ClusterIP                       
```

Reply:
For the external infra cluster scenario we have two users the external infra cluster admin and the normal user that want to create a cluster there, then my question is which one should decide when to install the kccm ? 
The fact that LoadBalancer services are being "redirected" to the infra load balancer is something to be decided by admin or by cluster creator ?

Reply:
oh interesting. i didn't realize clusterctl could do that. 

Starting with a template + flavor sounds like a reasonable path. However i don't know how maintainable it will be to have kccm integrated into the capk templates. For example, if rbac related to kccm changes, then we'd have to know to change that rbac in the capk repo... and it's possible certain fields change based on the kccm release.

If we source the kccm manifests from the corresponding kccm release, then they will always be accurate.

Reply:
Ok, I am going to reuse the golang code that calls kustomize to put a tool at capk that generates the kccm flavor template, then we can call that periodically or manually to generate a PR with a kccm update, we have do it similar at other projects.

Reply:
Looks like everything that is on the template will be force namespaced to the tenant namespace

https://github.com/kubernetes-sigs/cluster-api/blob/6b042456383aa18ba61fc5de5c41901dd7267c60/cmd/clusterctl/client/repository/template.go#L131

So we cannot put there the "kube-system" role binding needed for kccm :-(, maybe we can propose something at cluster-api to skip the fix